Unify bearing and target CameraUpdate

danesfeder opened the pull request:
Closes #394 

This PR seeks to update the `MapboxMap` camera `float` bearing and `LatLng` target at the same time.  It also introduces a `LocationLayerOption#refreshIntervalInMillis()` to allow customization of how often you'd like the camera to update.  

@LukasPaczos @tobrun if this approach seems okay, I can use the same runnable with the layer, so both the camera and layer updates will reflect the interval provided by the options. 

TODO:
- [x] Updated docs

danesfeder commented:
This may also address https://github.com/mapbox/mapbox-plugins-android/issues/567, as you'll be able to essentially disable interpolation by increasing the refresh interval.

LukasPaczos commented on plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java:
```diff
@@ -513,8 +513,12 @@ private void initialize() {
-    locationLayerCamera = new LocationLayerCamera(
-      mapView.getContext(), mapboxMap, cameraTrackingChangedListener, options, onCameraMoveInvalidateListener);
+
+    CameraUpdateProvider cameraUpdateProvider = new CameraUpdateProvider(mapboxMap,
```
This provider's queue needs to be cleared similarly to `StaleStateManager#onStop`.

LukasPaczos commented on plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/CameraUpdateProvider.java:
```diff
@@ -0,0 +1,56 @@
+  private CameraUpdate buildCameraUpdate() {
+    CameraPosition cameraPosition = mapboxMap.getCameraPosition();
+    CameraPosition.Builder cameraPositionBuilder = new CameraPosition.Builder();
+    if (bearing > 0 && bearing != cameraPosition.bearing) {
```
I think our current logic doesn't ensure that the target bearing is always 0-360.

danesfeder replied:
Does the shortest rotation util address this? https://github.com/mapbox/mapbox-plugins-android/blob/master/plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/Utils.java#L32-L40

LukasPaczos replied:
I don't think that's the case if the previous and current bearings are >360 or <-360, and when both are < 0. We probably should update the utils method to better normalize the bearing. Also, the bearing can be equal to 0.

LukasPaczos replied:
Additionally, we need to check here if we actually want to modify camera's bearing, otherwise in a mode like `CameraModer.TRACKING`, where we are tracking only target, we are unable to rotate the camera.

LukasPaczos commented on plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerOptions.java:
```diff
@@ -73,6 +73,9 @@
   private static final long STALE_STATE_DELAY_MS = 30000;
 
+  // TODO docs
+  private static final int REFRESH_INTERVAL_IN_MILLIS_DEFAULT = 1;
```
I think we can make the default refresh rate of 60FPS? That'd require a delay of 16.67 ms. Let's test this out though and decide more experimentally.

danesfeder replied:
Sounds good, I'll update the variable and we can test it out

danesfeder commented:
@LukasPaczos addressed your comments and asked some follow up questions.  Test for the camera provider / docs for the options were also added.  

Now working with both the camera and layer (here at 1/fps):
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/8434572/42399669-cfc5c66a-813c-11e8-8230-ab7d0e7fb9ce.gif)

LukasPaczos commented on plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/CameraUpdateProvider.java:
```diff
@@ -0,0 +1,64 @@
+    if (bearing > 0 && bearing != cameraPosition.bearing) {
+      cameraPositionBuilder.bearing(bearing);
+    }
+    if (target != null && !target.equals(cameraPosition.target)) {
```
Similarly, as in https://github.com/mapbox/mapbox-plugins-android/pull/573/files#r200948002, we need to check if we are actually tracking the position and not only bearing, otherwise, we won't be able to move the camera.

LukasPaczos replied:
Maybe we can move those checks out of [LocationLayerCamera](https://github.com/mapbox/mapbox-plugins-android/blob/master/plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerCamera.java#L120-L133) and in here?

LukasPaczos commented on plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/CameraUpdateProvider.java:
```diff
@@ -0,0 +1,64 @@
+import com.mapbox.mapboxsdk.geometry.LatLng;
+import com.mapbox.mapboxsdk.maps.MapboxMap;
+
+class CameraUpdateProvider implements RefreshPluginRunnable.RefreshPluginListener {
```
Generally, what about implementing the `RefreshPluginRunnable.RefreshPluginListener` in the `LocationLayerCamera`? That would simplify the checks needed for the animation and reflect the setup that we have with the `LocationLayer`.

zugaldia commented:
I'm gonna close this PR as per https://github.com/mapbox/mapbox-plugins-android/issues/684#issuecomment-433947880 the functionality provided by the Location Layer Plugin is now part of the core Maps SDK (there's no need to add additional plugins) and the plugin is now deprecated. If this ticket is still relevant for the new `LocationComponent`, please cut a new PR on https://github.com/mapbox/mapbox-gl-native. Thanks!
